src/Generating/EnderDragonFightStructuresGen.cpp

```diff
@@ -0,0 +1,282 @@
+
+#include "Globals.h"
+
+#include "EnderDragonFightStructuresGen.h"
+#include "../Chunk.h"
+#include "../Entities/EnderCrystal.h"
+#include "../WorldStorage/SchematicFileSerializer.h"
+
+cEnderDragonFightStructuresGen::cEnderDragonFightStructuresGen(int a_Seed, AString & a_TowerProperties, int a_Radius) :
+	m_Noise(a_Seed)
+{
+	// Loads the fountain schematic
+	if (!cSchematicFileSerializer::LoadFromSchematicFile(m_Fountain, AString("Prefabs") + cFile::GetPathSeparator() + "SinglePieceStructures" + cFile::GetPathSeparator() + "EndFountain.schematic"))
+	{
+		LOGWARNING("EnderDragonFightStructuresGen is missing it's end fountain prefab, please update your cuberite server files! There will be no end fountain!");
+	}
+	// Reads the given tower properties
+	const auto TowerPropertiesVector = StringSplitAndTrim(a_TowerProperties, ";");
+	for (const auto & TowerProperty : TowerPropertiesVector)
+	{
+		const auto TowerPropertyVector = StringSplitAndTrim(TowerProperty, "|");
+		if (TowerPropertyVector.size() != 3)
+		{
+			LOGWARNING("Got unmatching parameters on generating obsidian pillars: %s, Please use \"Height|Radius|HasCage\", ...", TowerProperty);
+			continue;
+		}
+		int Height = std::stoi(TowerPropertyVector[0]);
+		int Radius = std::stoi(TowerPropertyVector[1]);
+		bool HasCage;
+		if (NoCaseCompare(TowerPropertyVector[2], "true") == 0)
+		{
+			HasCage = true;
+		}
+		else if (NoCaseCompare(TowerPropertyVector[2], "false") == 0)
+		{
+			HasCage = false;
+		}
+		else
+		{
+			LOGWARNING("Got unknown value for boolean if the tower: %s should have a cage! %s", TowerProperty, TowerPropertyVector[2]);
+			continue;
+		}
+		m_TowerProperties.emplace_back(sTowerProperties(Height, Radius, HasCage));
+	}
+	// A random angle in radian
+	double Angle = m_Noise.IntNoise1D(a_Seed) * M_PI + M_PI;
+	// Generate Positions in a circle
+	for (int i = 0; i < static_cast<int>(m_TowerProperties.size()); i++)
+	{
+		auto TowerPos = Vector3i(static_cast<int>(FloorC(a_Radius * cos(Angle))), 0, static_cast<int>(FloorC(a_Radius * sin(Angle))));
+		auto ChunkX = static_cast<int>(FloorC(TowerPos.x / cChunkDef::Width));
+		if (TowerPos.x < 0)
+		{
+			ChunkX--;
+		}
+		auto ChunkZ = static_cast<int>(FloorC(TowerPos.z / cChunkDef::Width));
+		if (TowerPos.z < 0)
+		{
+			ChunkZ--;
+		}
+
+		m_TowerPos[cChunkCoords(ChunkX, ChunkZ)] = TowerPos;
+		Angle = fmod(Angle + (2.0f * M_PI / static_cast<double>(m_TowerProperties.size())), 2.0f * M_PI);
+	}
+}
+
+
+
+
+
+void cEnderDragonFightStructuresGen::GenFinish(cChunkDesc & a_ChunkDesc)
+{
+	// If not in the chunks to write
+	if ((a_ChunkDesc.GetChunkX() > 6) ||
+		(a_ChunkDesc.GetChunkX() < -6) ||
+		(a_ChunkDesc.GetChunkZ() > 6) ||
+		(a_ChunkDesc.GetChunkZ() < -6))
+	{
+		return;
+	}
+	// Places the exit portal
+	auto Coords = a_ChunkDesc.GetChunkCoords();
+	if (Coords == cChunkCoords({0, 0}))
+	{
+		a_ChunkDesc.WriteBlockArea(m_Fountain, -3, 62, -3, cBlockArea::msSpongePrint);
```

Comment:
Would these coordinates work when I use a portal with a different size?

Reply:
No

Reply:
This could be easily fixed. Whyever you would change the prefab. When the dragon dies it will spawn the portal. This spawning would know the size of the portal

Reply:
We could make this configurable. We just need to the the space where the exit portal should spawn

Reply:
If the endportal prefab is configurable it makes sense that the generator knows how to handle prefabs with different sizes right?

Reply:
Yeah of course - but we need to note down the size somewhere for the ender dragon spawn the exit portal

or we search the area araound (0, 62, 0) and spawn in the area around it till we hit bedrock. Although this will break the server if someone ever breaks the bedrock of the fountain in creative mode.

Another thing i noticed that the end portal regenerated if someone travels to the end. That should be done by us also.

that still might break the exit portal

Reply:
The obsidian floor at spawn also regenerates everytime something travels to the end. Isn't the portal always at exactly 0,y,0 ?

Reply:
The regenerating part is harder for us at the moment as the generator isn't build to regenerate specific structures in existing chunks.

Reply:
> The obsidian floor at spawn also regenerates everytime something travels to the end.
Oh yeah... that thing...

> Isn't the portal always at exactly 0,y,0 ?
It is